pruning-planting.py

```diff
@@ -0,0 +1,105 @@
+        Given a row from the trees dataset, find the planting segment that contains
+        it. Then return that segment id, planting year, and replacement species.
+        """
+        # Construct a shapely point from the tree data
+        pt = Point(tree["Longitude"], tree["Latitude"])
+
+        # Based on the description of the tree, select which df to search
+        segments = planting_median_segments if str(
+            tree["Location Description"]
+        ).lower() == "median" else planting_street_segments
+
+        # Find the row with the minimum distance from the point.
+        row = segments.distance(pt).idxmin()
+        segment = segments.iloc[row]
+        return segment["SEGMENT"]
+
+    trees["SEGMENT"] = trees.progress_apply(nearest_planting_segment, axis=1)
+    planting_segments = pd.concat([planting_street_segments, planting_median_segments])
+    trees = trees.merge(
+        planting_segments[["SEGMENT", "YEAR", "REPLACE"]], on="SEGMENT", how="left"
+    )
+    trees = trees.rename(
+        columns={"YEAR": "planting_year", "REPLACE": "replacement_species"}
+    )
+    return trees
+
+
+def pruning_for_trees(trees):
+    years = {"1718": "2017-2018", "1819": "2018-2019", "1920": "2019-2020"}
+    for year in years:
+        name_streets = f"data/pruning/pruning{year}_streets.shp"
+        name_medians = f"data/pruning/pruning{year}_medians.shp"
+        pruning = pd.concat([load_dataset(name_streets), load_dataset(name_medians)])
+        pruning[year] = years[year]
+
+        trees = trees.merge(pruning[["SEGMENT", year]], on="SEGMENT", how="left")
+
+    def collapse_years(tree):
+        for year in years:
+            if not pd.isnull(tree[year]):
+                return tree[year]
+
+        return tree[year]
+
+    trees["pruning_year"] = trees.progress_apply(collapse_years, axis=1)
+    trees = trees.drop([y for y in years], axis=1)
+
+    return trees
+
+
+if __name__ == "__main__":
+    # Load the trees dataset.
+    infile = "data/trees.csv"
+    outfile = "build/json/trees.json"
+    print(f"Loading data from {infile}")
+    trees = pd.read_csv(infile)
+    trees = trees.head(1000)
+    print("Joining planting data...")
+    trees = planting_for_trees(trees)
+    print("Joining pruning data...")
+    trees = pruning_for_trees(trees)
+    print(f"Saving file to {outfile}")
+    columns = {
+        "Tree ID": "id",
+        "Species ID": "eol_id",
+        "Latitude": "latitude",
+        "Longitude": "longitude",
+        "Height Group": "height_group",
+        "Name Common": "name_common",
+        "Name Botanical": "name_botanical",
+        "SEGMENT": "segment",
+    }
+    trees = trees.rename(columns=columns)
+    out_columns = [columns[c] for c in columns]
+    out_columns = out_columns + ["planting_year", "replacement_species", "pruning_year"]
+    tmp = trees[out_columns].to_json(orient="records")
+    with open(outfile, "w") as out:
+        out.write(json.dumps(json.loads(tmp), indent=2))
```

Comment:
I know this is still in-progress, but fwiw when I ran this it results in dropping the majority of the tree models. i.e. 35,000 tree models when in, and only like 13,000 or something came out (forget the specific number).

Reply:
I think this drop is bc the WIP only includes a portion of the trees (adding this note here so that I can keep track)